### PR TITLE
Quote and Trade Service Integration

### PR DESCRIPTION
**Pull request by nalancer08**
### Type of change

- [ ] Chore
- [X] Feature
- [ ] Bug fix

### Linked issue

- [X] Not tracked

### Why
Adding the Quote service (create) and Trade service (create) to handle the TradeFlow 

## Evidence

https://user-images.githubusercontent.com/7268597/176504855-c8475c67-eb75-452c-8eaf-874356617398.mov

### CODE REVIEW

File: SDKAndroid/src/main/java/app/cybrid/sdkandroid/components/quote/view/QuoteConfirmationModal.kt [hunk 1]
```diff
@@ -0,0 +1,533 @@
...
+        }
+    }
+}
+
+@Composable
+private fun QuoteConfirmationLoading(textID:Int) {
+
+    Box(
+        modifier = Modifier
+            .height(320.dp)
```
Comment (jc-cybrid) [inline SDKAndroid/src/main/java/app/cybrid/sdkandroid/components/quote/view/QuoteConfirmationModal.kt:126]: It's good to avoid magic numbers or constants spread all over the project and have them centralized for better maintenance. You could do that on a separate PR.

**Review by jc-cybrid — APPROVED** [review]
LGTM. Just a comment for future work.

File: SDKAndroid/src/main/java/app/cybrid/sdkandroid/components/quote/view/QuoteConfirmationModal.kt (hunk 1 above)
Comment (nalancer08) [reply SDKAndroid/src/main/java/app/cybrid/sdkandroid/components/quote/view/QuoteConfirmationModal.kt:126]: No, that is a bad design desition, into Compose (@Composable functions) the context suggest use Strings, Colors, Images, but dimensions no, cause the new APi standard use "dp" and "sp" and if we use values files we need to handle "context" or "ContextResource" and cast or transform any dimension in Composable dimensions

File: SDKAndroid/src/main/java/app/cybrid/sdkandroid/core/BigDecimal.kt [hunk 2]
```diff
@@ -43,10 +43,18 @@ class BigDecimal(internal val value:JavaBigDecimal) : Comparable<BigDecimal> {
         return BigDecimal(value.pow(n))
     }
 
+    fun toInt(): Int {
+        return value.toInt()
```
Comment (brent-cybrid) [inline SDKAndroid/src/main/java/app/cybrid/sdkandroid/core/BigDecimal.kt:47]: There will be loss of precision here. We should raise an exception when that occurs (same with Double).

**Review by brent-cybrid — COMMENTED** [review]
The tests you have are very very thin. They don't seem to test any edge or error cases.

File: SDKAndroid/src/main/java/app/cybrid/sdkandroid/core/BigDecimal.kt (hunk 2 above)
Comment (nalancer08) [reply SDKAndroid/src/main/java/app/cybrid/sdkandroid/core/BigDecimal.kt:47]: Yeah!! But let me explain what happen here!

Our class BigDecimal its a child of java.math.BigDecimal, so the methos "toInt()" and "toDouble()" are methos from the parent and this methods call the exception for us, like the error that i show you to fix the OpenGenerator was from this class, for that reason i dont raise any exeption.

I added some test raising the exeption!

File: SDKAndroid/src/main/java/app/cybrid/sdkandroid/core/BigDecimal.kt [hunk 3]
```diff
@@ -43,8 +43,26 @@ class BigDecimal(internal val value:JavaBigDecimal) : Comparable<BigDecimal> {
         return BigDecimal(value.pow(n))
     }
 
+    fun toInt(): Int {
+
+        val intValue = value.toInt()
+        if (intValue < 0) {
```
Comment (brent-cybrid) [inline SDKAndroid/src/main/java/app/cybrid/sdkandroid/core/BigDecimal.kt:49]: Can we compare `intValue == value.toInt()`?

Same below for double. That would be a better test.

File: SDKAndroid/src/test/java/app/cybrid/sdkandroid/components/quote/view/QuoteConfirmationModalTest.kt [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        // -- When
+        composeTestRule.setContent {
+            QuoteConfirmationLoading(R.string.trade_flow_quote_confirmation_modal_title)
+        }
+
+        // -- Then
+        composeTestRule.onNodeWithText("Order Quote").assertIsDisplayed()
+        composeTestRule.onNodeWithTag(Constants.QuoteConfirmation.LoadingIndicator.id).assertIsDisplayed()
+    }
+}
```
[the remark this answers is not in the document]
Comment (nalancer08) [reply SDKAndroid/src/test/java/app/cybrid/sdkandroid/components/quote/view/QuoteConfirmationModalTest.kt:71]: This specific case only check the UI Component its presented well in the screen.
The case of network fails its checked into TradeFlow test but if it fails the modal continue trying to get a correct order.

We dont have a flow in design if some error happens to show a error to the user. Good thing to implemented.

File: SDKAndroid/src/main/java/app/cybrid/sdkandroid/core/BigDecimal.kt (hunk 3 above)
Comment (nalancer08) [reply SDKAndroid/src/main/java/app/cybrid/sdkandroid/core/BigDecimal.kt]: Done, we compare with intValueExact, that allows to have the last correct number before Arithmetic throw raise!!

`intValue == value.intValueExact()`

**Review by brent-cybrid — COMMENTED** [review]
Looks good, but need to add a couple tests to get codecov to 100%

File: SDKAndroid/src/test/java/app/cybrid/sdkandroid/core/BigDecimalTest.kt [hunk 5]
```diff
@@ -192,6 +207,144 @@ class BigDecimalTest {
...
+        // -- When
+        try {
+            value = bigDecimalNumber.toDouble()
+        } catch (e: Throwable) {
+            error = e
+        }
+
+        // -- Then
+        Assert.assertNull(error?.cause)
+        Assert.assertNull(error?.message)
```
Comment (brent-cybrid) [inline SDKAndroid/src/test/java/app/cybrid/sdkandroid/core/BigDecimalTest.kt:242]: Should these be asserting null?

Comment (nalancer08) [reply SDKAndroid/src/test/java/app/cybrid/sdkandroid/core/BigDecimalTest.kt:242]: yes!! cause the cause and message are null, but the assert it not null

File: SDKAndroid/src/test/java/app/cybrid/sdkandroid/core/BigDecimalTest.kt [hunk 6]
```diff
@@ -192,6 +207,144 @@ class BigDecimalTest {
...
+
+        // -- Given
+        val shaNumber = "1157920892373161954235709850086879078532699846656405640394575"
+        val bigDecimalNumber = BigDecimal(shaNumber).setScale(100)
+
+        // -- When
+        val value = bigDecimalNumber.toDouble()
+
+        // -- Then
+        Assert.assertEquals(value, 1.23456789123456794E18, 1.0)
```
Comment (brent-cybrid) [inline SDKAndroid/src/test/java/app/cybrid/sdkandroid/core/BigDecimalTest.kt:221]: `Assert.assert(false)`

File: SDKAndroid/src/test/java/app/cybrid/sdkandroid/core/BigDecimalTest.kt (hunk 5 above)
Comment (brent-cybrid) [reply SDKAndroid/src/test/java/app/cybrid/sdkandroid/core/BigDecimalTest.kt:242]: `Assert.assertNotNull(error)`

File: SDKAndroid/src/test/java/app/cybrid/sdkandroid/core/BigDecimalTest.kt [hunk 7]
```diff
@@ -192,6 +207,144 @@ class BigDecimalTest {
...
+    fun toIntFailureTest() {
+
+        // -- Given
+        val bigDecimal1 = BigDecimal(1234567891234567891)
+
+        // -- When
+        val value = bigDecimal1.toInt()
+
+        // -- Then
+        Assert.assertEquals(value, -10714012610)
```
Comment (brent-cybrid) [inline SDKAndroid/src/test/java/app/cybrid/sdkandroid/core/BigDecimalTest.kt:290]: Replace this with an assert that always fails

**Review by brent-cybrid — APPROVED** [review]
Awesome job, @nalancer08 . Thanks for bearing with the reviews and making the changes :).
